Enable incremental build.

chm-tm opened the pull request:
Set the `CopyDocumentationFileToOutputDirectory` property to `false` and use both the `DocFileItem` and `FinalDocFile` items instead of an extra backup copy. This addresses issue #12.

chm-tm committed:
Enable incremental build.
The backup copy is no longer being produced. Instead, we take the C# compiler output from the intermediate directory and put the output where it is expected. This acts like a drop-in replacement for the original copy task which we now mute.

chm-tm commented:
Note that the `DocFileItem` and `FinalDocFile` items already existed back in the .NET Framework 2.0 days.

saucecontrol commented on src/InheritDoc/package/build/SauceControl.InheritDoc.targets:
```diff
@@ -3,24 +3,25 @@
+	<ItemGroup Condition=" '$(InheritDocEnabled)' == 'true' ">
+		<!-- Force intermediate documentation file -->
+		<DocFileItem Remove="$(DocumentationFile)"/>
+		<DocFileItem Include="$(IntermediateOutputPath)$(AssemblyName).xml"/>
```
This is not a good change.  The doc path may be set explicitly in the project file and may not be in the `obj` folder.

It is not guaranteed that `DocFileItem` and `FinalDocFile` are different paths, which is why `_InheritDocBackupFile` creates a unique name.

chm-tm replied:
We need to consider two kind of projects here:

1. Old projects explicitly specify `DocumentationFile`, typically in the `bin` folder.
In this case, `DocFileItem` and `FinalDocFile` coincide. Since `FinalDocFile` points to the `bin` folder due to its definition, we can guarantee different paths when we let `DocFileItem` point to the intermediate output path.
2. New projects specify `GenerateDocumentationFile` as `true` and the SDK lets `DocFileItem` point to the `obj` folder and `FinalDocFile` to the `bin` folder.
In this case, the redefinition of `DocFileItem` is a no-op.

I'm aware that this is a bit trickier than the `_InheritDocBackupFile` approach and might break customized builds more easily. For example, a customized build might have several items in `DocFileItem`.

I'd like to add some comments here.

saucecontrol replied:
Ok, yeah, that makes sense.  I think this should be conditional on `DocFileItem` being the same as `FinalDocFile` so that it only catches projects of the first type and doesn't unintentionally change anything in projects of the second type.

So far, I only see it set here
https://github.com/dotnet/msbuild/blob/23b938b931b48da647e9dbca42323ee60abf1126/src/Tasks/Microsoft.Common.CurrentVersion.targets#L400

saucecontrol reviewed (commented):
I appreciate the contribution, but neither your issue nor your PR description have enough words 😅

Can you explain in more detail what this is supposed to fix and how exactly it fixes that thing?

chm-tm commented:
> I appreciate the contribution, but neither your issue nor your PR description have enough words 😅

That happens to me quite often 🤷‍♂️.

> Can you explain in more detail what this is supposed to fix and

 I've now tried to clarify in issue #12 that I'd like to have a better build performance.

> how exactly it fixes that thing?

The way how the original `_InheritDocPostProcess` target specifies its inputs and outputs lets MSBuild execute it in consecutive builds. This is since the target input of the consecutive build is what the `InheritDocTask` formerly built, and this is newer than the backup file (declared as target output) even though the C# compiler didn't run in the consecutive build.

It's impossible to set up incremental build behavior when the documentation output of the C# compiler is the same as the output of the `InheritDocTask`. This is why this pull request resorts to another pair of inputs and outputs. It lets the `InheritDocTask` do the job of "copying" the `XML` to the output path. In order for this to work consistently, it tries to force that `DocFileItem` and `FinalDocFile` don't coincide. It does this by adjusting the `DocFileItem`.

saucecontrol commented:
> This is since the target input of the consecutive build is what the InheritDocTask formerly built, and this is newer than the backup file (declared as target output) even though the C# compiler didn't run in the consecutive build.

Ah, that's helpful, thanks.

> This is why this pull request resorts to another pair of inputs and outputs. It lets the InheritDocTask do the job of "copying" the XML to the output path.

Ok, so that's why you've disabled `CopyDocumentationFileToOutputDirectory` when `_InheritDocPostProcess` runs...

> In order for this to work consistently, it tries to force that DocFileItem and FinalDocFile don't coincide. It does this by adjusting the DocFileItem.

And that explains the rewrite of `DocFileItem`, though I would prefer it not change the name entirely as is done in the PR (forcing it to `$(AssemblyName).xml`).

I'm unclear how the current check for `_DocumentationFileProduced` is failing to prevent running when the C# compiler doesn't run.  I would hope that any changes we make in this area also address whatever was the cause of https://github.com/saucecontrol/InheritDoc/issues/10, so I'm trying to get a better understanding of the VS workflows (I personally don't run InheritDoc in VS, so it's been a while since I looked at the order of MSBuild tasks when running in VS).

I'll do some testing to make sure none of the scenarios that were previously working are affected and get back to you if I have further questions.

Thanks again for the extra explanation :)

chm-tm committed:
Only touch DocFileItem if required for incremental build to work.
Make the redefinition more generic by using the information from the FinalDocFile item instead of hardcoding the documentation name.

chm-tm committed:
Add more explanation on the intent of the authoring.
Adjust condition style to the rest of the file.

saucecontrol commented on src/InheritDoc/package/build/SauceControl.InheritDoc.targets:
```diff
@@ -3,24 +3,27 @@
-
 		<ItemGroup>
-			<FileWrites Condition="Exists('@(_InheritDocBackupFile)')" Include="@(_InheritDocBackupFile)" />
+			<FileWrites Condition="Exists('@(FinalDocFile)')" Include="@(FinalDocFile)" />
```
This will no longer work as intended, as it previously ran after the `Copy` task, which would have created the `_InheritDocBackupFile` it was conditionally adding to `FileWrites`.

Since the `FinalDocFile` is now created by `InheritdocTask`, the update to `FileWrites` needs to happen after that.

saucecontrol reviewed (changes requested):
I've gone back through the tangle of MSBuild `props` and `targets` involved in XML doc generation, and I believe your approach is sound.  I like this better than the backup file :)

One fix is needed for the `FileWrites` update, and I think we're good.  I'll do a bit more testing just to make sure I can't break it.

chm-tm committed:
Don't add the documentation output to the `FileWrites` item until the `InheritDocTask` has actually produced it.

chm-tm commented:
Thank you for taking the time to delve into the depths of MSBuild targets.

saucecontrol commented:
Thanks for the quick update.

I've been thinking through the possible side effects of changing `CopyDocumentationFileToOutputDirectory`, and there are a few possible negatives I've come up with.

1) If someone were to insert another target that runs after the package's .targets file and subsequently sets `InheritDocEnabled` to false, `InheritDocTask` won't run, but we'll also have disabled the copy from the `CopyFilesToOutputDirectory` task.  This one could be solved by setting `CopyDocumentationFileToOutputDirectory` to false inside `_InheritDocPostProcess`, when we know for sure it will run.
2) `CopyDocumentationFileToOutputDirectory` is relatively new-ish https://github.com/dotnet/msbuild/pull/3664, at least compared to the rest of the MSBuild properties/tasks we're hooking.  I'm not sure I care much about people running > ~5 year old versions of MSBuild, but the negative is that if someone is using an older version, that property will be ignored by `CopyFilesToOutputDirectory` and the unprocessed XML file from the intermediate output would overwrite the processed one.
3) What if someone set `CopyDocumentationFileToOutputDirectory` to false themselves because they didn't want the doc file in the output directory for whatever reason?  Now we're overriding their wish and copying it there anyway.  I don't know if that's realistic for users of InheritDoc, but it's a possible breaking change.

I think as far as solving the incremental build issue, your solution is very good, but I wonder if there's a way we can accomplish that without possible breaking changes.  Do you have any ideas?

chm-tm commented:
Regarding your points:
1. Can do that
2. I'd be sorry to completely break compatibility for a VS in the extended support period. I still have some instances running 2017 (time flies by fastly) and I know of a customer locked to a 2015 tool chain. What about running our target after `CopyFilesToOutputDirectory` target for old MSBuild versions?
3. What about checking `CopyDocumentationFileToOutputDirectory` for _not_ being set to `false` (the default sets it in the `CopyFilesToOutputDirectory` target? If it's false we could skip our target and issue a warning about this decision.

saucecontrol committed:
handle CopyDocumentationFileToOutputDirectory edge cases, reword comment

saucecontrol commented:
> users locked to legacy versions just needed to stay within the current major release of the Inheritdoc NuGet package. This would be acceptable from my point of view.

That was my thought as well.  I hate to drop support for something that was working if people still use it, but updating the minimum target allows for some cleanup, and most new features in newer versions of this package will be related to language features that are only usable with a modern SDK/VS.

> What about checking CopyDocumentationFileToOutputDirectory for not being set to false (the default sets it in the CopyFilesToOutputDirectory target? If it's false we could skip our target and issue a warning about this decision.

That sounds good to me.  It's an edge case that will likely never come up, but I'd rather warn and do nothing than silently do the wrong thing.

I've pushed an update if you don't mind giving it a look over to see if I missed anything.

I'll add an error message for older MSBuild versions in another commit later, along with some changes to the package metadata to keep older clients from taking the update.

chm-tm reviewed (commented):
LGTM

saucecontrol commented:
Thanks for fixing this and for the help while I got back up to speed with MSBuild 😄
